core-processor/src/ext.rs

```diff
@@ -432,33 +400,78 @@ impl Ext {
         }
     }
 
-    fn charge_gas(&mut self, val: u64) -> Result<(), ChargeError> {
-        let common_charge = self.context.gas_counter.charge(val);
-        let allowance_charge = self.context.gas_allowance_counter.charge(val);
-        self.check_charge_results(common_charge, allowance_charge)
-    }
-
     fn charge_sending_fee(&mut self, delay: u32) -> Result<(), ChargeError> {
         if delay == 0 {
-            self.charge_gas(self.context.message_context.settings().sending_fee())
+            self.charge_gas_if_enough(self.context.message_context.settings().sending_fee())
         } else {
-            self.charge_gas(
+            self.charge_gas_if_enough(
                 self.context
                     .message_context
                     .settings()
                     .scheduled_sending_fee(),
             )
         }
     }
+}
+
+impl CountersOwner for Ext {
+    fn charge_gas_runtime(&mut self, cost: RuntimeCosts) -> Result<(), ChargeError> {
+        let token = cost.token(&self.context.host_fn_weights);
+        let common_charge = self.context.gas_counter.charge_token(token);
+        let allowance_charge = self.context.gas_allowance_counter.charge_token(token);
+        self.check_charge_results(common_charge, allowance_charge)
+    }
+
+    fn charge_gas_runtime_if_enough(&mut self, cost: RuntimeCosts) -> Result<(), ChargeError> {
+        let token = cost.token(&self.context.host_fn_weights);
+        let common_charge = self.context.gas_counter.charge_token_if_enough(token);
+        let allowance_charge = self.context.gas_allowance_counter.charge_token(token);
+        self.check_charge_results(common_charge, allowance_charge)
+    }
 
-    fn refund_gas(&mut self, val: u64) -> Result<(), ChargeError> {
-        if self.context.gas_counter.refund(val) == ChargeResult::Enough {
-            self.context.gas_allowance_counter.refund(val);
+    fn charge_gas_if_enough(&mut self, amount: u64) -> Result<(), ChargeError> {
+        let common_charge = self.context.gas_counter.charge_if_enough(amount);
+        let allowance_charge = self.context.gas_allowance_counter.charge(amount);
+        self.check_charge_results(common_charge, allowance_charge)
+    }
+
+    fn refund_gas(&mut self, amount: u64) -> Result<(), ChargeError> {
+        if self.context.gas_counter.refund(amount) == ChargeResult::Enough {
+            self.context.gas_allowance_counter.refund(amount);
             Ok(())
         } else {
             Err(ChargeError::TooManyGasAdded)
         }
     }
+
+    fn gas_left(&self) -> GasLeft {
+        GasLeft {
+            gas: self.context.gas_counter.left(),
+            allowance: self.context.gas_allowance_counter.left(),
+        }
+    }
+
+    fn set_gas_left(&mut self, gas_left: GasLeft) {
```

Comment:
here you just drop results from calls `charge_if_enough` so I'd go with unwraps with unreachable calls as checked within the fn ops

Reply:
Done